software/apps/romi_control/states.c

```diff
@@ -404,101 +307,76 @@ outputs_t substate_transition(inputs_t input_state, system_state_t* curr_state){
 			output.left_speed = -speed;
 			output.right_speed = speed;
 			curr_state->substate.substate = ROTATING;
-			break;
 		}
+		break;
 	}
 	case AVOIDANCE: {
 				
 		float diff_left = curr_state->substate.avoidance_distance - curr_state->substate.total_distance_traveled_left;
 		float diff_right = curr_state->substate.avoidance_distance - curr_state->substate.total_distance_traveled_right;
 		
 		float avg_diff = (diff_left + diff_right) / 2;
-		if(input_state.bump_left || input_state.bump_right){
+		if(input_state->bump_left || input_state->bump_right){
 
-			curr_state->substate.most_recent_bump = input_state.bump_left ? LEFT_BUMP : RIGHT_BUMP;
+			curr_state->substate.most_recent_bump = input_state->bump_left ? LEFT_BUMP : RIGHT_BUMP;
 			curr_state->substate.substate = STOPPED;
-			break;
 		} else if(fabs(avg_diff) < distance_threshold){
 			
 			curr_state->substate.turn_angle_substate =  angle_modulo(atan2f(-curr_state->position_y, curr_state->distance_to_travel - curr_state->position_x) * 180 / M_PI - curr_state->curr_orientation_angle);
 			curr_state->substate.next_state_turning = FORWARD;
 			curr_state->substate.substate = ROTATING;
 		} else {
-		
-			float wheel_diff = diff_right - diff_left;
-			int8_t sign_left = (2 * (diff_left > 0)) - 1;
-			int8_t sign_right = (2 * (diff_right > 0)) - 1;
-			
-			float dist_traveled_left = measure_distance(input_state.left_encoder, curr_state->substate.previous_left_encoder);
-			float dist_traveled_right = measure_distance(input_state.right_encoder, curr_state->substate.previous_right_encoder);
-			curr_state->substate.total_distance_traveled_left += dist_traveled_left;
-			curr_state->substate.total_distance_traveled_right += dist_traveled_right;
-
-			float avg_dist = (dist_traveled_right + dist_traveled_left) / 2;
-			curr_state->position_x += cosf(curr_state->curr_orientation_angle * M_PI /  180) * avg_dist;
-			curr_state->position_y += sinf(curr_state->curr_orientation_angle * M_PI / 180) * avg_dist;
-
-			curr_state->substate.previous_left_encoder = input_state.left_encoder;
-			curr_state->substate.previous_right_encoder = input_state.right_encoder;
-			output.left_speed = sign_left * fmax(fabs(k_dist * diff_left - k_diff * wheel_diff), min_drive_speed);
-			output.right_speed = sign_right * fmax(fabs(k_dist * diff_right + k_diff * wheel_diff), min_drive_speed);
-			break;
+			translation_control(curr_state, input_state, 1, &(output.left_speed), &(output.right_speed));
 		}
-	}
-	default:
 		break;
 	}
+	}
 	if (curr_state->substate.substate != old_state) {
-     		substate_transition_out(input_state, &(curr_state->substate), old_state);
+     		substate_transition_out(&(curr_state->substate), old_state);
      		substate_transition_in(input_state, &(curr_state->substate));
     }
 	return output;
 }
+
+void translation_control(system_state_t* curr_state, inputs_t* input_state, int dir, int16_t* left_pwr, int16_t* right_pwr) {
```

Comment:
maybe explicitly set `dir` to be an `int8_t`

Reply:
Is there a benefit beyond reducing the size of the variable by a couple bytes? Wherever it's being used would still require  a conversion of the value from floats to int16_t